### ISSUE TITLE
Connection to MeshAgent gets stuck on "Setup..."

### ISSUE THREAD

**Issue by whalehub** [Ylianst/MeshCentral#32]
At some point during the past two weeks, the agent control features of MeshCentral stopped working for me. Whenever I click on the "Connect" button on the "Desktop", "Terminal" or "Files" tab, the connection gets stuck on "Setup..." indefinitely. I'm positive that it's not related to any change that was made in MeshCentral because I tried rolling back to the previous 10 versions (all of which had previously been working fine) and reinstalled MeshAgent and MeshCentral from scratch with no success.

MeshAgent is installed on a Windows 10 machine with the latest updates installed. I've tried disabling the firewall and anti-virus software and uninstalling recent Windows updates, but it didn't solve the issue for me.

The "Console" tab in MeshCentral only shows this output:

`New tunnel connection #1: wss://my.domain.tld:443/meshrelay.ashx?id=<random string>, rights: 4294967295`

Has anyone else experienced this issue? Any ideas on how to fix it?

**Edit:** Added more information.

**Comment by Ylianst** [comment]
Hi. Sorry for the delay. Best way to debug this, run MeshCentral from the command prompt with "--debug 1". Once the agents reconnect, try to do a remote desktop connect, you should see something like this:

```
Relay holding: zpcd44llm3 (::1) Authenticated
Relay connected: zpcd44llm3 (192.168.2.139 --> ::1)
Relay disconnect: zpcd44llm3 (::1 --> 192.168.2.139)
```
The first line says "Authenticated", this is the browser connecting to the server correctly. The second line is the agent connecting. I then do some work and disconnect on the 3rd line. You will be able to see if the problem is on the browser or agent side. If it's a browser problem, try a different browser and look in the browser console for any errors.

This should help see what is going on.
Ylian

**Comment by whalehub** [comment]
@Ylianst Thanks for the tip. I just ran MeshCentral with the debug flag and saw this in the logs:

Relay holding: nkg0nqceg2 (<IP>) Authenticated

The second and third line don't appear in the logs. This means that the problem is on the agent side, right? Any idea what could be blocking the connection all of the sudden? It used to work flawlessly with the same firewall and anti-virus setup that I have now.

**Comment by Ylianst** [comment]
Correct, the agent's remote desktop session is not connecting to the server. I have no idea what the problem could be. Are you using any firewalls / reverse proxies or interesting network setup? If using a reverse proxy, which one? I will check about giving you more debug commands.

**Comment by Ylianst** [comment]
Ha! We figured it out. If you are using a reverse proxy with many certificates, the MeshAgent will try to connect the tunnel without TLS-SNI and encounter the wrong certificate and not connect right. We are working on a fix for this. Right now, only the main agent connection has SNI, not the tunnels.

**Comment by Ylianst** [comment]
Just published MeshCentral v0.2.3-w that should fix the agent tunnel connection problem when the server is also using a reverse proxy. Update and post back if it works or not. Thanks.

**Comment by whalehub** [comment]
@Ylianst Thanks a lot for the fix, that did the trick. I'm really happy that it's working again now. :-)
